api/lib/OCR.js

```diff
@@ -0,0 +1,67 @@
+const fs = require('fs');
+const path = require('path');
+const exec = require('child_process').exec;
+
+const gs = require('ghostscript4js');
+const tesseract = require('node-tesseract');
+
+const errorMessages = require('./error-messages');
+
+function OCR(options) {
+    if (typeof options === "object") {
+        const acceptedKeys = Object.keys(this.options);
+        Object.keys(options).forEach((val) => {
+            if (!acceptedKeys.includes(val)) return delete options[val];
+            if (typeof this.options[val] !== typeof options[val]) {
+                throw Error(errorMessages.TYPE_MISMATCH_ERROR(val, typeof this.options[val], typeof options[val]));
+            }
+        });
+        this.options = Object.assign(this.__proto__.options, options);
+    }
+}
+
+OCR.prototype.options = {
+    l: 'por',
+    psm: 6,
+    binary: '/usr/bin/tesseract'
+};
+
+OCR.prototype.getText = (file) => {
+    file = path.dirname(require.main.filename) + '/' + file;
+    if (path.extname(file) !== '.pdf')
+        throw Error(errorMessages.FILE_TYPE_MISMATCH_ERROR('pdf', path.extname(file)));
+    return new Promise((res, rej) => {
+        PDFtoImage(file).then((data) => {
+            let text = '';
+            let count = 0;
+            data.map((image, index) => {
+                tesseract.process(image, this.options, function (err, result) {
+                    if (err) return rej(err);
+                    text = text + '\n===================\n' + result;
+                    fs.unlink(image, (err) => { if (err) console.log(err); });
+                    count++;
+                    if (count === data.length) return res(text);
+                });
+            })
+        }).catch((err) => {
+            rej(err);
+        })
+    });
```

Comment:
`then` returns a promise that follows these rules:

After the invocation of the handler function, if the handler function:

 - returns a value, the promise returned by `then` gets resolved with the returned value as its value;
 - throws an error, the promise returned by `then` gets rejected with the thrown error as its value;
 - returns an already resolved promise, the promise returned by `then` gets resolved with that promise's value as its value;
 - returns an already rejected promise, the promise returned by `then` gets rejected with that promise's value as its value.
 - returns another pending promise object, the resolution/rejection of the promise returned by `then` will be subsequent to the resolution/rejection of the promise returned by the handler. Also, the value of the promise returned by `then` will be the same as the value of the promise returned by the handler.

Reply:
Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then

Reply:
Btw, this implementation still has the problem of page ordering, the final OCR may have the PDF text in the wrong page order.

I thought on a solution for this, as follows:

 - Create a function that receives a string (to hold the final OCR), an array and an index
 - at the end of the execution check if index < array.length
 - if positive, return a call to itself passing the ocr string, the array and index+1
- return the final string otherwise

This way, we guarantee that page 2 only executes after page 1 finishes, and so on.
This shouldn't break the promise, theoretically, `then` would wait for everything to end processing before resolving

Let me know if you have any questions